src/Language/ASKEE/Exposure/Interpreter.hs

```diff
@@ -743,6 +747,104 @@ runSim (SimDiffEq delta) t mdl =
+interpretModelSkillRank modelObsSamples alphas truth =
+  do modelObsSamples' <- array (array (array double)) modelObsSamples
+     alphas'          <- array double alphas
+     truth'           <- array double truth
+     pure $ VArray (VDouble <$> skillRank modelObsSamples' alphas' truth')
+
+-- The first parameter is a list of samples indexed by (model number,
+-- observation number, sample number). The second param is a list of alphas
+-- Third param is the "ground truth", i.e. a list of values indexed by
+-- observation number. Returns a list of ranks (value in [0,1]) indexed by model
+-- number (higher is better)
+skillRank :: [[[Double]]] -> [Double] -> [Double] -> [Double]
+skillRank modelObservations alphas truth =
+  -- Returns the _mean_ of the ranks across all observations for a given model
+  [ mean (Map.findWithDefault [0.0] mi ranksByModel)
+  | mi <- [0..n-1]
+  ]
+  where
+    -- Group all of the per-observation ranks by model.
+    ranksByModel =
+      foldr (\(mi, r) -> Map.insertWith (++) mi [r]) mempty rankedByObservation
+
+    -- Given all scores (indexed by model and observation number), For each
+    -- observation assign each model a rank, (lower rank means lower interval
+    -- score for that observation)
+    rankedByObservation =
+      assignRanks =<< byObservation
+
+    byObservation =
+        groupBy ((==)`on`obsNo)
+      $ sortBy (compare`on`obsNo) wmi
+
+    assignRanks :: [(Int, Int, Double)] -> [(Int, Double)]
+    assignRanks obsScores =
+      [ (modelNo s, rank (fromIntegral r))
+      | s <- sortBy (compare`on`score) obsScores
+      | r <- [(1::Int)..]
+      ]
+
+    rank :: Double -> Double
+    rank r = 1.0 - ((r - 1.0)/(fromIntegral n - 1))
+
+    n = length modelObservations
+
+    -- Calculate WIS for all models mi and all observations
+    wmi = concat [ goModel mi m | m <- modelObservations | mi <- [0..] ]
+
+    obsNo (_,i,_)    = i
+    modelNo (mi,_,_) = mi
+    score (_,_,s)    = s
+
+    -- for model mi:
+    -- for each observation i
+    -- calculate the WIS at i.
+    goModel :: Int -> [[Double]] -> [(Int, Int, Double)]
+    goModel mi obsSamples =
+      [ (mi, i, wis obs alphas y) | obs <- obsSamples | y <- truth | i <- [0..] ]
+
+
+-- | Weighted interval score for a given set of samples, alphas and observation.
+wis :: [Double] -> [Double] -> Double -> Double
+wis vals alphas y =
+  (1/(k + 0.5)) * (w0 * abs (y-m) + agg_interval)
+  where
+    k = fromIntegral (length alphas - 1)
+    w0 = 0.5
+    go a = (a/2) -- w_k
+         * isAlpha (quant (a/2)) (quant (1 - (a/2))) a y -- interval score
+    agg_interval = sum (go <$> alphas)
+
+    m     = quant 0.5
+    quant = quantile vals'
+    vals' = sort vals
+
+-- The first argument should be sorted in increasing order
+quantile :: [a] -> Double -> a
+quantile sortedData q = sortedData !! safe
+  where
+    safe   = max 0 (min needed (length sortedData - 1))
+    needed = floorDoubleInt (q * fromIntegral (length sortedData))
```

Comment:
If I'm reading this right, the use of `floorDoubleInt` means that if a quantile lies between two data points, then it will pick the lower of the two data points. (In other words, the behavior of [`interpolation=lower` in `numpy.quantile`](https://numpy.org/doc/stable/reference/generated/numpy.quantile.html)). Is that correct? If so, it would be worth mentioning this in a comment.

Reply:
Yes that's right, but I'm still not sure if that's the behavior we actually want? I suspect it doesn't matter too much.

Reply:
Right, I think the only thing that's really important is that all quantile measurements agree on the method of interpolation (which is obviously the case here).